### ISSUE TITLE
cannot receive: failed to read from stream

### ISSUE THREAD

**Issue by bytec77** [psy0rz/zfs_autobackup#72]
hi, 

I´m testing zfs_autobackup for the first time but I run into error

```bash
root@pve:~# zfs-autobackup --keep-source=2 --keep-target=5 --ssh-source 10.0.1.220 prox220 rpool/backup/prox220 --progress --verbose --debug
  zfs-autobackup v3.0 - Copyright 2020 E.H.Eefting (edwin@datux.nl)
  
  #### Source settings
  [Source] Datasets on: 10.0.1.220
  [Source] Keep the last 2 snapshots.
  [Source] Selects all datasets that have property 'autobackup:prox220=true' (or childs of datasets that have 'autobackup:prox220=child')
  
  #### Selecting
# [Source] Getting selected datasets
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'get' '-t' 'volume,filesystem' '-o' 'name,value,source' '-s' 'local,inherited' '-H' 'autobackup:prox220'
  [Source] rpool: Ignored (disabled)
  [Source] rpool/ROOT: Ignored (disabled)
  [Source] rpool/ROOT/pve-1: Ignored (disabled)
  [Source] rpool/data: Ignored (disabled)
  [Source] rpool/data/basevol-110-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-100-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-101-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-104-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-105-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-106-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-107-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-108-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-109-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-112-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-113-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-114-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-115-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-116-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-117-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-118-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-119-disk-0: Ignored (disabled)
  [Source] rpool/data/subvol-999-disk-0: Selected (direct selection)
  [Source] rpool/data/vm-102-disk-0: Ignored (disabled)
  [Source] rpool/data/vm-103-disk-0: Ignored (disabled)
  [Source] rpool/data/vm-111-disk-0: Ignored (disabled)
  [Source] rpool/data/vm-200-disk-0: Ignored (disabled)
  [Source] rpool/data/vm-201-disk-0: Ignored (disabled)
  [Source] rpool/data/vm-202-disk-0: Ignored (disabled)
  [Source] rpool/data/vm-202-disk-1: Ignored (disabled)
  [Source] rpool/data/vm-203-disk-1: Ignored (disabled)
  [Source] rpool/test: Ignored (disabled)
  
  #### Snapshotting
# [Source] rpool/data/subvol-999-disk-0: Getting snapshots
# [Source] rpool/data/subvol-999-disk-0: Checking if filesystem exists
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'list' 'rpool/data/subvol-999-disk-0'
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'list' '-d' '1' '-r' '-t' 'snapshot' '-H' '-o' 'name' 'rpool/data/subvol-999-disk-0'
# [Source] rpool/data/subvol-999-disk-0: Getting bytes written since our last snapshot
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'get' '-H' '-ovalue' '-p' 'written@rpool/data/subvol-999-disk-0@prox220-20210405172322' 'rpool/data/subvol-999-disk-0'
  [Source] rpool/data/subvol-999-disk-0: No changes since prox220-20210405172322
  [Source] No changes anywhere: not creating snapshots.
  
  #### Target settings
  [Target] Datasets are local
  [Target] Keep the last 5 snapshots.
  [Target] Receive datasets under: rpool/backup/prox220
  
  #### Sending and thinning
# [Target] rpool/backup/prox220: Checking if filesystem exists
# [Target] RUN    >  zfs list rpool/backup/prox220
# [Target] rpool/backup/prox220/rpool/data: Checking if filesystem exists
# [Target] RUN    >  zfs list rpool/backup/prox220/rpool/data
# [Source] zpool rpool: Getting zpool properties
# [Source] RUN    >  ssh 10.0.1.220 'zpool' 'get' '-H' '-p' 'all' 'rpool'
# [Target] zpool rpool: Getting zpool properties
# [Target] RUN    >  zpool get -H -p all rpool
# [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0: Determining start snapshot
# [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0: Getting snapshots
# [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0: Checking if filesystem exists
# [Target] RUN    >  zfs list rpool/backup/prox220/rpool/data/subvol-999-disk-0
# [Target] STDERR > cannot open 'rpool/backup/prox220/rpool/data/subvol-999-disk-0': dataset does not exist
# [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0: Creating virtual target snapshots
# [Source] rpool/data/subvol-999-disk-0: Create thinning list
# [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0: Getting zfs properties
# [Source] rpool/data/subvol-999-disk-0: Getting zfs properties
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'get' '-H' '-o' 'property,value' '-p' 'all' 'rpool/data/subvol-999-disk-0'
# [Source] rpool/data/subvol-999-disk-0@prox220-20210405170907: Transfer snapshot to rpool/backup/prox220/rpool/data/subvol-999-disk-0
  [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0@prox220-20210405170907: receiving full
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'send' '-L' 'zfs_autobackup_option_test'
# [Source] STDERR > cannot open 'zfs_autobackup_option_test': dataset does not exist
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'send' '-e' 'zfs_autobackup_option_test'
# [Source] STDERR > cannot open 'zfs_autobackup_option_test': dataset does not exist
# [Source] RUN    >  ssh 10.0.1.220 'zfs' 'send' '-c' 'zfs_autobackup_option_test'
# [Source] STDERR > cannot open 'zfs_autobackup_option_test': dataset does not exist
# [Source] PIPE   >  ssh 10.0.1.220 'zfs' 'send' '-L' '-e' '-c' '-v' '-P' '-p' 'rpool/data/subvol-999-disk-0@prox220-20210405170907' |
# [Target] RUN    >  zfs recv -s zfs_autobackup_option_test
# [Target] STDERR > cannot receive: failed to read from stream
# [Target] rpool/backup/prox220/rpool/data/subvol-999-disk-0@prox220-20210405170907: Enabled resume support
# [Target] RUN    >  zfs recv -u -v -s rpool/backup/prox220/rpool/data/subvol-999-disk-0
# [Target] Piping input
! [Target] STDERR|> warning: cannot send 'rpool/data/subvol-999-disk-0@prox220-20210405170907': Unknown error 1030
! [Target] STDERR > cannot receive: failed to read from stream
! [Source] rpool/data/subvol-999-disk-0: FAILED: Command '(pipe)' returned non-zero exit status 1.
! Exception: Command '(pipe)' returned non-zero exit status 1.
Traceback (most recent call last):
  File "/usr/local/bin/zfs-autobackup", line 1862, in <module>
    sys.exit(zfs_autobackup.run())
  File "/usr/local/bin/zfs-autobackup", line 1825, in run
    fail_count=self.sync_datasets(source_node, source_datasets)
  File "/usr/local/bin/zfs-autobackup", line 1707, in sync_datasets
    source_dataset.sync_snapshots(target_dataset, show_progress=self.args.progress, features=common_features, filter_properties=filter_properties, set_properties=set_properties, ignore_recv_exit_code=self.args.ignore_transfer_errors, source_holds= not self.args.no_holds, rollback=self.args.rollback, raw=self.args.raw, other_snapshots=self.args.other_snapshots, no_send=self.args.no_send, destroy_incompatible=self.args.destroy_incompatible)
  File "/usr/local/bin/zfs-autobackup", line 1295, in sync_snapshots
    source_snapshot.transfer_snapshot(target_snapshot, features=features, prev_snapshot=prev_source_snapshot, show_progress=show_progress,   filter_properties=allowed_filter_properties, set_properties=allowed_set_properties, ignore_recv_exit_code=ignore_recv_exit_code, resume_token=resume_token, raw=raw)
  File "/usr/local/bin/zfs-autobackup", line 1060, in transfer_snapshot
    target_snapshot.recv_pipe(pipe, features=features, filter_properties=filter_properties, set_properties=set_properties, ignore_exit_code=ignore_recv_exit_code)
  File "/usr/local/bin/zfs-autobackup", line 1022, in recv_pipe
    self.zfs_node.run(cmd, input=pipe, valid_exitcodes=valid_exitcodes)
  File "/usr/local/bin/zfs-autobackup", line 453, in run
    raise(subprocess.CalledProcessError(input.returncode, "(pipe)"))
subprocess.CalledProcessError: Command '(pipe)' returned non-zero exit status 1.
```

**Comment by psy0rz** [comment]
Its an actual error returned by zfs send:
```
! [Target] STDERR|> warning: cannot send 'rpool/data/subvol-999-disk-0@prox220-20210405170907': Unknown error 1030
```

I'll do a quick google to see what turns up.

What operating system and version is running on the source system? 
Is it up to date?
Does zpool status report any errors?
Does dmesg show any errors related to the disk or zfs kernel modules?

Edwin

**Comment by psy0rz** [comment]
Hmm i cant find anything usefull on google..

**Comment by bytec77** [comment]
The zfs version of the pools were different so I make a system upgrade and an zpool upgrade -a on both sides and now it seems to work. 

Now I have one warning and I don't know how bad this issue is. 

```bash
[Target] STDERR|> Warning: incremental source didn't specify type, assuming snapshot. Use '@' or '#' prefix to avoid ambiguity.
```

**Comment by psy0rz** [comment]
Ah thats great to hear. :)

Yep you can ignore that error, its fixed in version 3.1. (try the beta, its great :)

Edwin
